pkg/reconciler/apibinding/apibinding_controller.go

```diff
@@ -0,0 +1,348 @@
+	apiResourceSchemaLister  apislisters.APIResourceSchemaLister
+	apiResourceSchemaIndexer cache.Indexer
+	crdLister                apiextensionslisters.CustomResourceDefinitionLister
+	crdIndexer               cache.Indexer
+
+	deletedCRDTracker *lockedStringSet
+}
+
+// enqueueAPIBinding enqueues an APIBinding .
+func (c *controller) enqueueAPIBinding(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.Infof("Queueing APIBinding %q", key)
+	c.queue.Add(key)
+}
+
+// enqueueAPIExport enqueues maps an APIExport to APIBindings for enqueuing.
+func (c *controller) enqueueAPIExport(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	klog.Infof("Mapping APIExport %q", key)
+	list, err := c.apiBindingsIndexer.ByIndex(indexAPIBindingsByWorkspaceExport, key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, apiBinding := range list {
+		c.enqueueAPIBinding(apiBinding)
+	}
+}
+
+// enqueueCRD maps a CRD to APIResourceSchema for enqueuing.
+func (c *controller) enqueueCRD(obj interface{}) {
+	crd, ok := obj.(*apiextensionsv1.CustomResourceDefinition)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("obj is supposed to be a CustomResourceDefinition, but is %T", obj))
+		return
+	}
+
+	if crd.Annotations[annotationSchemaClusterKey] == "" || crd.Annotations[annotationSchemaNameKey] == "" {
+		return
+	}
+
+	apiResourceSchemaKey := clusters.ToClusterAwareKey(crd.Annotations[annotationSchemaClusterKey], crd.Annotations[annotationSchemaNameKey])
+	apiResourceSchema, err := c.apiResourceSchemaLister.Get(apiResourceSchemaKey)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	c.enqueueAPIResourceSchema(apiResourceSchema)
+}
+
+// enqueueAPIResourceSchema maps an APIResourceSchema to APIExports for enqueuing.
+func (c *controller) enqueueAPIResourceSchema(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	apiExports, err := c.apiExportsIndexer.ByIndex(indexAPIExportsByAPIResourceSchema, key)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+
+	for _, apiExport := range apiExports {
+		c.enqueueAPIExport(apiExport)
+	}
+}
+
+// Start starts the controller, which stops when ctx.Done() is closed.
+func (c *controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s controller", controllerName)
+	defer klog.Infof("Shutting down %s controller", controllerName)
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(ctx, c.startWorker, time.Second)
+	}
+
+	<-ctx.Done()
+}
+
+func (c *controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *controller) processNextWorkItem(ctx context.Context) bool {
```

Comment:
(No action required) We have a lot of these functions and they are largely the same. Maybe a shared helper is suggested (not in this PR)?

Reply:
which would bring us into controller runtime land. The *_controller.go files are mostly standard. We decided to not overabstract these. Maintenance is low, if you follow a good example to cargo cult.

Reply:
I'm not sure I buy that. N controllers all duplicating the same code is definitely a smell...